### PR TITLE
Added example #2 serial chat

### PR DESCRIPTION
**Pull request by amellalalex**
Example 2 is fresh off the press.

Building on what was explored in the previous example, simple serial port I/O is expanded into multithreaded and asynchronous exchanges between a master and slave pseudo-terminal device pair.

At its core, C code is interacting with the low-level system calls to handle the pseudo-terminal serial ports. Then, higher level C++ structures wrap the underlying C code to improve code readability and simplify usage.

The objective of this example was to better simulate how serial ports may be used within the context of a satellite and a ground station.

Let me know any feedback you might have regarding the code!

### CODE REVIEW

File: Linux Serial Ports/2. Serial chat/README.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+`$ ./chat join <slave endpoint path>`
+* Opens the slave serial device's endpoint (aka joining the chat room).
+* Prompts user for input and sends it to the master.
+* Listens for incoming data from the master and prints it to the screen.
+
+## Prompts
+Prompts are fronted with custom specifiers, described as follows:
+* `[*]`     Signifies a generic prompt.
+* `[!]`     Signifies an error message.
+* `[(!)]`   Signifies a non-critical error message (execution continues anyway).
```
Comment (liamhebert) [inline Linux Serial Ports/2. Serial chat/README.md:46]: You may be interested in looking into the `spdlog` library. Built for logging and has many utilities such as different log levels (like here!) and text file cycling. Either way, great idea

File: Linux Serial Ports/2. Serial chat/include/SerialPorts.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
+#ifndef SERIAL_PORTS_H
+#define SERIAL_PORTS_H
+
+// Required for pseudoterminal functions to work
+#define _XOPEN_SOURCE 1
+#define _GNU_SOURCE   1
+
+////////////////////////////
+//    Relevant headers    //
+////////////////////////////
```
Comment (liamhebert) [inline Linux Serial Ports/2. Serial chat/include/SerialPorts.h:10]: Such large comments are not really necessary, but nice anyway :)

File: Linux Serial Ports/2. Serial chat/include/SerialPorts.hpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+namespace SerialPortsCPP
+{
+  class MasterPort
+  {
+  public:
+    string slave;
+    thread *transmitter = NULL;
+    thread *listener    = NULL;
+
+    void Open();
```
Comment (liamhebert) [inline Linux Serial Ports/2. Serial chat/include/SerialPorts.hpp:23]: Make sure you are consistent in your casing. In `SerialPorts.h`, you are writing in lower_case versus here where you use Capital casing.

File: Linux Serial Ports/2. Serial chat/src/cpp_wrapper.cpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  void MasterPort::Open()
+  {
+    // Create master pseudo-terminal device
+    if(master_create() != 0)
+    {
+      throw *(new PortOpenFailed);
+    }
+  }
+
+  void MasterPort::Listen()
```
Comment (liamhebert) [inline Linux Serial Ports/2. Serial chat/src/cpp_wrapper.cpp:18]: hmm... both Listen() and Transmit() don't really do what you expect them to do as they only instantiate the threads. It almost seems like these should be in some sort of constructor or renamed to `StartListening` or `StartTransmitting`. Maybe a more verbose name here would help, but they do seem like constructor type methods

File: Linux Serial Ports/2. Serial chat/src/cpp_wrapper.cpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  void SlavePort::Open(string path)
+  {
+    // Open slave device at specified endpoint
+    if(slave_create(path.c_str()) != 0)
+    {
+      throw *(new PortOpenFailed);
+    }
+  }
+
+  void SlavePort::Listen()
```
Comment (liamhebert) [inline Linux Serial Ports/2. Serial chat/src/cpp_wrapper.cpp:56]: Ditto here for above

File: Linux Serial Ports/2. Serial chat/src/main.c [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+{
+  // Check arguments
+  if(argc < 2)
+  {
+    printf("[!] Not enough arguments\n");
+    exit(1);
+  }
+
+  // Master (create chat room)
+  if(strcmp(argv[1], "create") == 0)
```
Comment (liamhebert) [inline Linux Serial Ports/2. Serial chat/src/main.c:13]: This master code could be relegated to its own executable for code cleanness. Minor concern only though, this works fine as an example. In our final product however, these should be separate

File: Linux Serial Ports/2. Serial chat/src/main.cpp [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    if(string(argv[1]) == "create")
+    {
+      // Create master serial port object
+      MasterPort master;
+
+      // Create master/slave pseudo-terminal pair.
+      master.Open();
+
+      // Start listening/transmitting to slave
+      master.Listen();
```
Comment (liamhebert) [inline Linux Serial Ports/2. Serial chat/src/main.cpp:35]: I already kinda touched on this, but I feel like this should be combined into `Start()` or part of the constructor

File: Linux Serial Ports/2. Serial chat/src/master.c [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+  // Receive messages until quit is received
+  while(strcmp(buffer, "quit\n") != 0)
+  {
+    int read_bytes = -1;
+
+    // Read message from serial port
+    if((read_bytes = read(port, buffer, buff_len)) < 0)
+    {
+      printf("[(!)] Failed to read message from slave.\n");
+      exit(1);
```
Comment (liamhebert) [inline Linux Serial Ports/2. Serial chat/src/master.c:70]: I feel like this should be `pthread_exit` instead

**Review by liamhebert — APPROVED** [review]
Good stuff Alex, this is a really good example of IPC. Going forward, I definitely recommend moving away from maintaining both a C++ and C version of the code. While I personally think that keeping just the C++ version would be ideal, it really depends on what you are most comfortable with. Most of my comments here are related to naming conventions, it really looks great. We should meet up sometime and get a roadmap to launch established and start working on concrete mission code :). Besides that, this is really looking great

**Review by hermit-shell — APPROVED** [review]
thanks

**Pull request merged**